cg.go

```diff
@@ -0,0 +1,162 @@
+package optimize
+
+import (
+	"math"
+
+	"github.com/gonum/floats"
+)
+
+const (
+	iterationRestartFactor = 6
+	angleRestartTreshold   = 0.9
+)
+
+type CGVariant interface {
```

Comment:
I don't particularly like the word Variant. I understand what you mean, but it's a pretty vague term. Wikipedia calls it a "Preconditioner" which is much more descriptive. If you don't think that's misleading, I'd rather see CGPreconditioner (or CGPrecon) than CGVariant.